### PR TITLE
m_clones: Show a list over nicks under each CIDR

### PR DESCRIPTION
**Pull request by ShutterQuick**
closes #736

### CODE REVIEW

File: src/modules/m_clones.cpp [hunk 1]
```diff
@@ -64,10 +151,23 @@ class CommandClones : public Command
...
 	CommandClones cmd;
+
  public:
-	ModuleClones() : cmd(this)
+	ModuleClones() : cmd(this, utrack)
+	{
+		utrack.SetTimeout(60*5);
+	}
+
+	void OnPostConnect(User*user) CXX11_OVERRIDE
```
Comment (attilamolnar) [inline src/modules/m_clones.cpp]: Missing space between type and var name

File: src/modules/m_clones.cpp [hunk 2]
```diff
@@ -23,16 +23,97 @@
...
+
+		if (timeout)
+		{
+			tracking = true;
+			SetInterval(timeout);
+		}
+
+		for (user_hash::const_iterator it = ServerInstance->Users->clientlist.begin(); it != ServerInstance->Users->clientlist.end(); ++it)
+		{
+			User* user = &(*it->second);
```
Comment (attilamolnar) [inline src/modules/m_clones.cpp]: `it->second`

File: src/modules/m_clones.cpp [hunk 3]
```diff
@@ -23,16 +23,97 @@
 
 /** Handle /CLONES
  */
```
Comment (attilamolnar) [inline src/modules/m_clones.cpp]: This comment belongs to the CommandClones class, should put the new code above it

File: src/modules/m_clones.cpp [hunk 4]
```diff
@@ -23,16 +23,97 @@
...
+		tracking = false;
+		return false;
+	}
+
+	void SetTimeout(unsigned long t)
+	{
+		timeout = t;
+	}
+
+	CloneList& GetMap()
```
Comment (attilamolnar) [inline src/modules/m_clones.cpp]: Couldn't it return a const ref?

File: src/modules/m_clones.cpp [hunk 5]
```diff
@@ -23,16 +23,97 @@
...
- 	CommandClones(Module* Creator) : Command(Creator,"CLONES", 1)
+	CommandClones(Module* Creator, UserTracker& ut)
+		: Command(Creator,"CLONES", 1), utrack(ut)
 	{
 		flags_needed = 'o'; syntax = "<limit>";
 	}
 
 	CmdResult Handle (const std::vector<std::string> &parameters, User *user)
 	{
+		CloneList& clonemap = utrack.GetMap();
```
Comment (attilamolnar) [inline src/modules/m_clones.cpp]: Move this near first use (that is, before the loop like the current code does with the clonemap)

File: src/modules/m_clones.cpp [hunk 6]
```diff
@@ -23,16 +23,97 @@
 
 /** Handle /CLONES
  */
+
+
+typedef TR1NS::unordered_set<User*> UList;
+typedef std::map<irc::sockets::cidr_mask, UList > CloneList;
```
Comment (attilamolnar) [inline src/modules/m_clones.cpp]: Extra space after UList and `>`

File: src/modules/m_clones.cpp [hunk 7]
```diff
@@ -21,19 +21,112 @@
...
+	void DelClient(User* user)
+	{
+		if (!tracking)
+			return;
+
+		CloneList::iterator i = cl.find(user->GetCIDRMask());
+		if (i == cl.end())
+		{
+			ServerInstance->Logs->Log(MODNAME, LOG_DEFAULT, "Potential bug. CIDR was not found in clone map upon removal. Did you change clone CIDR?");
+			return StopTracking();
```
Comment (attilamolnar) [inline src/modules/m_clones.cpp]: StopTracking();
    return;

File: src/modules/m_clones.cpp [hunk 8]
```diff
@@ -21,19 +21,112 @@
...
+		StopTracking();
+		return false;
+	}
+
+	void SetTimeout(unsigned long t)
+	{
+		timeout = t;
+	}
+
+	const CloneList& GetMap()
```
Comment (attilamolnar) [inline src/modules/m_clones.cpp]: Method now can be const

File: src/modules/m_clones.cpp [hunk 9]
```diff
@@ -21,19 +21,112 @@
...
+
+		UList& counts = i->second;
+		UList::iterator it = std::find(counts.begin(), counts.end(), user);
+		if (it == counts.end())
+		{
+			ServerInstance->Logs->Log(MODNAME, LOG_DEFAULT, "Potential bug. User was not found in clone map upon removal. Did you change clone CIDR?");
+			return StopTracking();
+		}
+
+		stdalgo::vector::swaperase(counts, it);
```
Comment (attilamolnar) [inline src/modules/m_clones.cpp]: The above if should check the return of this too, if it returns false the user was not found and the same should happen

**Comment by SadieCat** [comment]
The clones module was moved to -contrib for v4.

**Pull request closed without merging**